cmd/snap-bootstrap/cmd_initramfs_mounts.go

```diff
@@ -269,114 +285,690 @@ func copyFromGlobHelper(src, dst, globEx string) error {
+		} else {
+			m.degradedState.partition("ubuntu-boot").LocateState = partitionNotFound
+			m.degradedState.LogErrorf("error locating ubuntu-boot partition on disk %s: %v", m.disk.Dev(), err)
+		}
+
+		return m.unlockDataFallbackKey, nil
 	}
 
+	dev := fmt.Sprintf("/dev/disk/by-partuuid/%s", partUUID)
+	m.degradedState.partition("ubuntu-boot").Device = dev
+	m.degradedState.partition("ubuntu-boot").LocateState = partitionFound
+
 	// should we fsck ubuntu-boot? probably yes because on some platforms
 	// (u-boot for example) ubuntu-boot is vfat and it could have been unmounted
 	// dirtily, and we need to fsck it to ensure it is mounted safely before
 	// reading keys from it
 	fsckSystemdOpts := &systemdMountOptions{
 		NeedsFsck: true,
 	}
-	if err := doSystemdMount(fmt.Sprintf("/dev/disk/by-partuuid/%s", partUUID), boot.InitramfsUbuntuBootDir, fsckSystemdOpts); err != nil {
-		return err
+	if err := doSystemdMount(dev, boot.InitramfsUbuntuBootDir, fsckSystemdOpts); err != nil {
+		// didn't manage to mount boot, so try to use fallback key for data
+		m.degradedState.LogErrorf("failed to mount ubuntu-boot: %v", err)
+		m.degradedState.partition("ubuntu-boot").MountState = partitionErrMounting
+		return m.unlockDataFallbackKey, nil
+	}
+
+	// verify ubuntu-boot comes from same disk as ubuntu-seed
+	if err := m.verifyMountPoint(boot.InitramfsUbuntuBootDir, "ubuntu-boot"); err != nil {
+		return nil, err
 	}
 
-	// 2.X+1, verify ubuntu-boot comes from same disk as ubuntu-seed
-	matches, err := disk.MountPointIsFromDisk(boot.InitramfsUbuntuBootDir, nil)
+	m.degradedState.partition("ubuntu-boot").MountState = partitionMounted
+	m.degradedState.partition("ubuntu-boot").MountLocation = boot.InitramfsUbuntuBootDir
+
+	// next step try to unlock data with run object
+	return m.unlockDataRunKey, nil
+}
+
+// stateUnlockDataRunKey will try to unlock ubuntu-data with the normal run-mode
+// key, and if it fails, progresses to the next state, which is either:
+// - failed to unlock data, but we know it's an encrypted device -> try to unlock with fallback key
+// - failed to find data at all -> try to unlock save
+// - unlocked data with run key -> mount data
+func (m *stateMachine) unlockDataRunKey() (stateFunc, error) {
+	// TODO: don't allow recovery key at all for this invocation, we only allow
+	// recovery key to be used after we try the fallback key
+	runModeKey := filepath.Join(boot.InitramfsBootEncryptionKeyDir, "ubuntu-data.sealed-key")
+	unlockOpts := &secboot.UnlockVolumeUsingSealedKeyOptions{
+		// don't allow using the recovery key to unlock, we only try using the
+		// recovery key after we first try the fallback object
+		AllowRecoveryKey: false,
+		// don't lock keys, we manually do that at the end always, we don't know
+		// if this call to unlock a volume will be the last one or not
+		LockKeysOnFinish: false,
+	}
+	unlockRes, err := secbootUnlockVolumeUsingSealedKeyIfEncrypted(m.disk, "ubuntu-data", runModeKey, unlockOpts)
```

Comment:
This is something to consider later, but the more I look at the handling of the returned value and error the more I think secboot.UnlockVolumeUsingSealedKeyIfEncrypted is doing too much instead of leaving the finding of the partition to the caller.